int32[] dir1
int32[] dir2
int32[] dir3
int32[] dir4